api/apiclient.go

```diff
@@ -329,7 +237,19 @@ func PerferredHost(info *Info) string {
 func loginWithContext(ctx context.Context, st *state, info *Info) error {
 	result := make(chan error, 1)
 	go func() {
-		result <- st.Login(info.Tag, info.Password, info.Nonce, info.Macaroons)
+		// if loginProvider is not specified we use the default login provider
+		// which logs in using username/password or macaroons.
+		if st.loginProvider == nil {
+			result <- errors.Errorf("login provider not specified")
+		}
+
+		loginResult, err := st.loginProvider.Login(ctx, st)
+		if err != nil {
+			result <- err
+			return
+		}
```

Comment:
I think we could improve this - we're using `st` to get the provider and then also as a parameter to Login. At the time we parse the account details, can we set up st so that we just need to then call `st.Login(info)` or something. In fact, I can't see where the info passed into loginWithContext is used.

So spitballing, something like (not necessarily this exactly)

```
func loginWithContext(ctx context.Context, st *conn, info *Info) error {
	result := make(chan error, 1)
	go func() {
		result <- st.Login(ctx, info)
	}()
	select {
	case err := <-result:
		return errors.Trace(err)
	case <-ctx.Done():
		return errors.Annotatef(ctx.Err(), "cannot log in")
	}
}
```
The above keeps the login provider abstract hidden from the login flow. And we aren't cherry picking attributes off info to pass in (which are provider dependent).

Reply:
@wallyworld ptal - changed: st does not need to hold the loginProvider as it only holds the state once login completes.. also Info is not needed in the loginWithContext call..